clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -3783,17 +3801,6 @@ raw2trace_t::raw2trace_t(
     decode_cache_.reserve(cache_count);
     for (int i = 0; i < cache_count; ++i)
         decode_cache_.emplace_back(cache_count);
-
-#if defined(AARCH64)
-    // TODO i#6556, i#1684: The decoder uses a global sve_veclen variable to store the
-    // vector length value it uses when decoding. drdecodelib ends up being linked into
-    // drcachesim twice: once into the drcachesim executable, and one into libdynamorio.
-    // When we call dr_standalone_init() above it will initialize the version of
-    // sve_veclen in libdynamorio, but not the one in drcachesim.
-    // Unfortunately it is the version of sve_veclen in drcachesim that gets used when
-    // decoding in raw2trace so we need to explicitly initialize its sve_veclen here.
-    dr_set_sve_vector_length(proc_get_vector_length_bytes() * 8);
```

Comment:
If someone processes an older trace without the new marker, no one will set the length and it will be the default which seems to be 0?  If 0, what does the decoder do: does it error out?  Should there be a better default?

Reply:
Outside of drmemtrace, this seems to make using DR's standalone decoder painful for SVE: if the default is 0 would someone decoding SVE have the decoder fall over?  Anyone using the decoder now has to call the set vector length function?  Should that be documented?

Reply:
> If someone processes an older trace without the new marker, no one will set the length and it will be the default which seems to be 0? If 0, what does the decoder do: does it error out? Should there be a better default?

Currently, if the vector length is set to 0 the decoder will not produce an error. It will just use 0 in its calculations which will result in some SVE load/store instructions being decoded with a 0 byte offset when they shouldn't be. drdisas has a `-vl` command line option to set the vector length. We could add an equivalent option to drcachesim to allow users to specify a vector length for traces which don't contain a vector length marker.

> Outside of drmemtrace, this seems to make using DR's standalone decoder painful for SVE: if the default is 0 would someone decoding SVE have the decoder fall over? Anyone using the decoder now has to call the set vector length function? Should that be documented?

In order to handle vector length changes properly in DynamoRIO (#6625) I think we are going to have to change the way vector length is stored and managed, so that might mean changes to `dr_set_sve_vector_length()`. We can make sure as part of that issue that how it works, and any requirements on standalone decoder users are properly documented.

Reply:
Probably the default should be 128 or something valid so that reasonable results are obtained without explicit calls to set vector lengths: I suspect 99% of users are not going to make that call on the first pass and a reasonable default should be provided.

Reply:
> so that reasonable results are obtained without explicit calls to set vector lengths:

Explicit failures may be better than hidden trace/decoding correctness issues.

> a reasonable default should be provided.

Instead of a fixed number, could we initialize it using proc_get_vector_length_bytes in the decoder? E.g., https://github.com/DynamoRIO/dynamorio/blob/62d441e8bd0c8aad489e8a02dcb421db30d72da4/core/ir/opnd_shared.c#L2762

If it cannot be obtained from the environment, it may be better to fail, so that the user can set it explicitly. (except in cases where we know a certain number to be a very good default)

Reply:
Having a decoder that fails when you try to use if you don't make a call to set the vector length does not sound user-friendly to me: most will not know they're supposed to call that; many will not care and just want to see the opcode in offline decoding.  Some kind of general initialization routine seems better; default also seem good to me.

Let's file a new issue on this.  It relates to several other issues on trying to make drdecode work without explicit initialization, which maybe we have to abandon for other reasons.  I will file one and put in the xrefs.

Reply:
Filed #6646